src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,480 @@
+            // processing this one, it won't queue another request.
+            const request = this.queuedRequests[0];
+            try {
+                const result = await this.queryKeyBackup(request.roomId, request.megolmSessionId);
+
+                if (this.stopped) {
+                    return;
+                }
+                // We got the encrypted key from backup, let's try to decrypt and import it.
+                try {
+                    await this.decryptAndImport(request, result);
+                } catch (e) {
+                    this.logger.error(
+                        `Error while decrypting and importing key backup for session ${request.megolmSessionId}`,
+                        e,
+                    );
+                }
+                // now remove the request from the queue as we've processed it.
+                this.queuedRequests.shift();
+            } catch (err) {
+                if (err instanceof KeyDownloadError) {
+                    switch (err.code) {
+                        case KeyDownloadErrorCode.MISSING_DECRYPTION_KEY:
+                            this.markAsNotFoundInBackup(request.megolmSessionId);
+                            // continue for next one
+                            this.queuedRequests.shift();
+                            break;
+                        case KeyDownloadErrorCode.NETWORK_ERROR:
+                            // We don't want to hammer if there is a problem, so wait a bit.
+                            await sleep(this.backoffDuration);
+                            break;
+                        case KeyDownloadErrorCode.STOPPED:
+                            // If the downloader was stopped, we don't want to retry.
+                            this.pauseLoop();
+                            return;
+                        case KeyDownloadErrorCode.CONFIGURATION_ERROR:
+                            // Backup is not configured correctly, so stop the loop.
+                            this.pauseLoop();
+                            return;
+                    }
+                } else if (err instanceof KeyDownloadRateLimit) {
+                    // we want to retry after the backoff time
+                    await sleep(err.retryMillis);
+                }
+            }
+        }
+        // all pending request have been processed, we can stop the loop.
+        this.pauseLoop();
+    }
+
+    /**
+     * Query the backup for a key.
+     *
+     * @param targetRoomId - ID of the room that the session is used in.
+     * @param targetSessionId - ID of the session for which to check backup.
+     */
+    private async queryKeyBackup(targetRoomId: string, targetSessionId: string): Promise<KeyBackupSession> {
+        const configuration = await this.getOrCreateBackupConfiguration(false);
+        if (!configuration) {
+            throw new KeyDownloadError(KeyDownloadErrorCode.CONFIGURATION_ERROR);
+        }
```

Comment:
it might be better to make the call to `getOrCreateBackupConfiguration` in `downloadKeysLoop` (and then pass the `backupVersion` into `queryKeyBackup`, and the `decryptor` into `decryptAndImport`), which would help ensure we have a matching pair of backup version and decryptor.

I don't insist though. YMMV.

Reply:
Yes good point, updated